### PR TITLE
Add slice access

### PR DESCRIPTION
**Pull request by willglynn**
This PR adds slice access to certain storages. Slices allow the user to bolt on an accelerator framework like OpenCL (#553) for high throughput component processing. My thoughts behind this particular design are https://github.com/slide-rs/specs/issues/553#issuecomment-525099211 and https://github.com/slide-rs/specs/issues/553#issuecomment-525362663.

## Checklist

* [x] I've added tests for all code changes and additions (where applicable)
* [x] I've added a demonstration of the new feature to one or more examples
* [x] I've updated the book to reflect my changes
* [x] Usage of new public items is shown in the API docs

## API changes

Non-breaking, strictly additive:

* Add `DenseVecStorage<T>::as_slice() -> &[T]` and `as_mut_slice() -> &mut [T]`.
* Add `VecStorage::unsafe_slice() -> &[T]` and `unsafe_mut_slice() -> &mut [T]`.
* Add `DefaultVecStorage<T>`, which works like `VecStorage<T>`, requires `T: Default`, and provides `as_slice() -> &[T]` and `as_mut_slice() -> &mut [T]`.
* Add `as` or `unsafe` `[_mut]_slice` directly on `Storage` as appropriate.

### CODE REVIEW

**Comment by WaDelma** [comment]
I think those `unsafe_slice`s should return raw pointers as this is constructing partially initialized slice which I think is UB. Not 100% sure about this though.

**Comment by WaDelma** [comment]
Well if the whole `Vec` is full then it would be safe yes. But it's practically useless. I would provide way of getting raw pointers and let the user construct slice out of it if they want.

**Comment by WaDelma** [comment]
If you have `VecStorage<bool>` that contains uninitialised values, this would create reference to slice of booleans that have different values than `0` or `1` and I think having such reference might be UB. But I cannot be certain as there is no memory model yet. What I am certain that having `&bool` to uninitialized bool is insta UB (and so &T of uninitialised value is easily UB).

https://www.reddit.com/r/rust/comments/95vxdy/understanding_ub_with_stdmemuninitialized/

https://github.com/rust-lang/rust/issues/53491

**Comment by willglynn** [comment]
I rearranged my changes and force-pushed the branch for this PR. The first commit adds `as_slice()`/`as_mut_slice()` to `DenseVecStorage` and adds `DefaultVecStorage`. This is sufficient for my use (and likely others' use) per #553. 

The second commit reworks `VecStorage` to use `MaybeUninit` per my previous comment and makes it expose `as_slice() -> &[MaybeUninit<T>]`. These changes are related (it would be _nice_ to slice `VecStorage`), but they are separable, and I'm happy to move that commit to a new PR if desired.

**Comment by WaDelma** [comment]
Note that for f32 every bit pattern isn't allowed: Signaling NaN is a UB.

I am not sure if slice is just syntactic sugar from compilers POV. But yeah probably constructing slice isn't UB.

I think having `MaybeUninit` is good direction to go to as it gives more confidence on the correctness of this stuff. I am not against bumping the minimum rust version myself, but cannot say if that's for others.

**Comment by willglynn** [comment]
> Note that for f32 every bit pattern isn't allowed: Signaling NaN is a UB.

🤔 Do you mean making such an `f32`, dereferencing such an `&f32`, or performing operations with such an `f32`?

[`f32::from_bits(v: u32)`](https://doc.rust-lang.org/std/primitive.f32.html#method.from_bits) says it's equivalent to `mem::transmute::<u32, f32>(v)`, and it mentions some platform-specific gotchas around NaN, but that's implementation-defined behavior and not undefined behavior.

> I think having `MaybeUninit` is good direction to go to as it gives more confidence on the correctness of this stuff. I am not against bumping the minimum rust version myself, but cannot say if that's for others.

:+1:

**Comment by WaDelma** [comment]
Related: https://github.com/slide-rs/specs/issues/646

**Comment by willglynn** [comment]
I note in https://github.com/slide-rs/specs/issues/646#issuecomment-536056356 why I now believe that slices must always refer to initialized elements in current Rust, contrary to my earlier comments here.

**Comment by WaDelma** [comment]
This PR looks otherwise good, but `DenseVecStorage` should probably be changed to also have `MaybeUninit`.

**Comment by willglynn** [comment]
Ah, the `Vec<T>`s is dense and always initialized, it's the the `Vec<Index>`es which are not. On it.

File: src/storage/storages.rs [hunk 1]
```diff
@@ -83,12 +96,40 @@ unsafe impl<T> DistinctStorage for HashMapStorage<T> {}
...
+/// cannot be compared with indices from any other storage, and
+/// a particular entity's position within this slice may change
+/// over time.
 #[derive(Derivative)]
 #[derivative(Default(bound = ""))]
 pub struct DenseVecStorage<T> {
     data: Vec<T>,
-    entity_id: Vec<Index>,
-    data_id: Vec<Index>,
+    entity_id: Vec<MaybeUninit<Index>>,
```
Comment (WaDelma) [inline src/storage/storages.rs]: Does `entity_id` need to contain `MaybeUninit`? We don't `set_len` on it, right?

Comment (willglynn) [reply src/storage/storages.rs]: Oops, got over-zealous.

**Pull request merged**